drivers/can/can_renesas_ra.c

```diff
@@ -0,0 +1,1171 @@
+
+	return clock_control_get_rate(cfg->custom.dll_clk, dll_subsys, rate) ? -EIO : 0;
+}
+
+/**
+ * Get maximum number of RX filters.
+ */
+static int can_renesas_ra_get_max_filters(const struct device *dev, bool ide)
+{
+	ARG_UNUSED(ide);
+	const struct can_renesas_ra_cfg *cfg = dev->config;
+
+	return cfg->custom.rx_filter_num;
+}
+
+#if CONFIG_CAN_FD_MODE
+/**
+ * Configure the bus timing for the data phase of a CAN FD controller.
+ */
+static int can_renesas_ra_set_timing_data(const struct device *dev,
+					  const struct can_timing *timing_data)
+{
+	struct can_renesas_ra_data *data = dev->data;
+
+	if (data->common.started) {
+		/* CAN controller is not in stopped state. */
+		return -EBUSY;
+	}
+
+	can_bit_timing_cfg_t data_timing = ZEPHYR_TO_FSP_TIMING(timing_data);
+
+	if (!r_canfd_bit_timing_parameter_check(&data_timing, true)) {
+		return -ENOTSUP;
+	}
+	data->custom.data_timing = data_timing;
+
+	return 0;
+}
+#endif /* CONFIG_CAN_FD_MODE */
+
+/**************************************************************************************************
+ * Internal function definition
+ *************************************************************************************************/
+void can_renesas_ra_fsp_cb(can_callback_args_t *p_args)
+{
+	const struct device *dev = p_args->p_context;
+
+	switch (p_args->event) {
+	case CAN_EVENT_RX_COMPLETE: {
+		struct can_frame z_frame = {0};
+
+		FSP_TO_ZEPHYR_FRAME(&p_args->frame, &z_frame);
+		can_renesas_ra_call_rx_cb(dev, &z_frame);
+		break;
+	}
+
+	case CAN_EVENT_TX_COMPLETE: {
+		can_renesas_ra_call_tx_cb(dev);
+		break;
+	}
+
+	case CAN_EVENT_ERR_CHANNEL: {
+		if (p_args->error & R_CANFD_CFDC_ERFL_BEF_Msk) {
+			can_renesas_ra_call_state_change_cb(dev, CAN_STATE_ERROR_ACTIVE);
+		}
+		if (p_args->error & R_CANFD_CFDC_ERFL_EWF_Msk) {
+			can_renesas_ra_call_state_change_cb(dev, CAN_STATE_ERROR_WARNING);
+		}
+		if (p_args->error & R_CANFD_CFDC_ERFL_EPF_Msk) {
+			can_renesas_ra_call_state_change_cb(dev, CAN_STATE_ERROR_PASSIVE);
+		}
+		if (p_args->error & R_CANFD_CFDC_ERFL_BOEF_Msk) {
+			can_renesas_ra_call_state_change_cb(dev, CAN_STATE_BUS_OFF);
+		}
+		break;
+	}
+
+	default:
+		break;
+	}
+}
+
+static inline void can_renesas_ra_call_tx_cb(const struct device *dev)
+{
+	struct can_renesas_ra_data *data = dev->data;
+
+	data->custom.tx_cb(dev, 0, data->custom.tx_usr_data);
+	k_work_cancel_delayable(&data->custom.tx_timeout_work);
+}
+
+static inline void can_renesas_ra_call_rx_cb(const struct device *dev, struct can_frame *rx_frame)
+{
+	struct can_renesas_ra_data *data = dev->data;
+	const struct can_renesas_ra_cfg *cfg = dev->config;
+	struct can_renesas_ra_filter *rx_filter = data->custom.rx_filter;
+
+	for (uint32_t i = 0; i < cfg->custom.rx_filter_num; i++) {
+		if (!rx_filter->rx_cb) {
+			rx_filter++;
+			continue; /* this filter has not set yet */
+		}
+
+		if (!can_frame_matches_filter(rx_frame, &rx_filter->filter)) {
+			rx_filter++;
+			continue; /* filter did not match */
+		}
```

Comment:
Please consider using lookup tables for the installed/matching hardware filter. Software-filtering like this is _very_ heavy on the CPU and kind of defeats the whole idea of HW filtering.

Reply:
Using lookup tables in this case is difficult now. Currently, the RX interrupt cannot know which filter entry the message came from, only the filter rule. It will be improved later, when the RX mailbox support interrupt and each filter entry will point to 1 mailbox.